controllers/tcproute_controller.go

```diff
@@ -0,0 +1,231 @@
+		}
+
+		//Check if referred gateway has the at least one listener with properties defined from TCPRoute parentref.
+		if err := r.verifyListener(ctx, gw, parentRef); err != nil {
+			log.Info("No matching listener found for referred gateway", "GatewayName", parentRef.Name, "GatewayPort", parentRef.Port)
+			//Check next parent ref.
+			continue
+		}
+
+		supportedGateways = append(supportedGateways, *gw)
+	}
+
+	//only used for logging
+	if len(supportedGateways) < 1 {
+		return false, nil, parentRefErr
+	}
+
+	refferedGateway := &supportedGateways[0]
+	log.Info("TCP Route appeared referring to Gateway", "Gateway ", refferedGateway.Name, "GatewayClass Name", refferedGateway.Spec.GatewayClassName)
+
+	return true, refferedGateway, nil
+}
+
+// verifyListener verifies that the provided gateway has at least one listener
+// matching the provided ParentReference.
+func (r *TCPRouteReconciler) verifyListener(ctx context.Context, gw *gatewayv1beta1.Gateway, tcprouteSpec gatewayv1alpha2.ParentReference) error {
+	for _, listener := range gw.Spec.Listeners {
+		if (listener.Protocol == gatewayv1beta1.TCPProtocolType) && (listener.Port == gatewayv1beta1.PortNumber(*tcprouteSpec.Port)) {
+			return nil
+		}
+	}
+	return fmt.Errorf("No matching Gateway listener found for defined Parentref")
+}
+
+func (r *TCPRouteReconciler) ensureTCPRouteConfiguredInDataPlane(ctx context.Context, tcproute *gatewayv1alpha2.TCPRoute, gateway *gatewayv1beta1.Gateway) error {
+	// build the dataplane configuration from the TCPRoute and its Gateway
+	targets, err := dataplane.CompileTCPRouteToDataPlaneBackend(ctx, r.Client, tcproute, gateway)
+	if err != nil {
+		return err
+	}
+
+	// TODO: add multiple endpoint support https://github.com/Kong/blixt/issues/46
+	dataplaneClient, err := dataplane.NewDataPlaneClient(context.Background(), r.Client)
+	if err != nil {
+		return err
+	}
+
+	confirmation, err := dataplaneClient.Update(context.Background(), targets)
+	if err != nil {
+		return err
+	}
+
+	r.log.Info(fmt.Sprintf("successful data-plane UPDATE, confirmation: %s", confirmation.String()))
+
+	return nil
+}
+
+func (r *TCPRouteReconciler) ensureTCPRouteDeletedInDataPlane(ctx context.Context, tcproute *gatewayv1alpha2.TCPRoute, gateway *gatewayv1beta1.Gateway) error {
+	// build the dataplane configuration from the TCPRoute and its Gateway
+	targets, err := dataplane.CompileTCPRouteToDataPlaneBackend(ctx, r.Client, tcproute, gateway)
+	if err != nil {
+		return err
+	}
+
+	// TODO: add multiple endpoint support https://github.com/Kong/blixt/issues/46
+	dataplaneClient, err := dataplane.NewDataPlaneClient(context.Background(), r.Client)
+	if err != nil {
+		return err
+	}
+
+	// delete the target from the dataplane
+	confirmation, err := dataplaneClient.Delete(context.Background(), targets.Vip)
+	if err != nil {
+		return err
+	}
+
+	r.log.Info(fmt.Sprintf("successful data-plane DELETE, confirmation: %s", confirmation.String()))
+
+	oldFinalizers := tcproute.GetFinalizers()
```

Comment:
WDYT about moving the finalizer removal out of this function? The function is named `ensureTCPRouteDeletedInDataPlane`, and removing also the dataplane finalizers looks a bit out of context.

Reply:
I think it's entirely relevant, the entire purpose of the finalizer is to hold up the object until we've confirmed removal from the dataplane.

Reply:
I think it's entirely relevant, the entire purpose of the finalizer is to hold up the object until we've confirmed removal from the dataplane. Keep in mind too that now that we're moving this project to Kubernetes SIGs it is intended to become a reference implementation for Gateway API, and as such it's going to undergo a LOT of refactoring and reviews from Gateway API community members before we start advertising it as a reference implementation (in fact, the intention right now is that before we consider it "ready" as a reference implementation we will do a full PR review of the code with the org).